go-controller/pkg/clustermanager/secondary_network_cluster_manager.go

```diff
@@ -0,0 +1,139 @@
+	nad "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/network-attach-def-controller"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// secondaryNetworkClusterManager object manages the multi net-attach-def controllers.
+// It implements networkAttachDefController.NetworkControllerManager and can be used
+// by NetAttachDefinitionController to add and delete NADs.
+type secondaryNetworkClusterManager struct {
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *nad.NetAttachDefinitionController
+	client        clientset.Interface
+	ovnClient     *util.OVNClusterManagerClientset
+	kube          kube.Interface
+	watchFactory  *factory.WatchFactory
+	stopChan      chan struct{}
+}
+
+func newSecondaryNetworkClusterManager(ovnClient *util.OVNClusterManagerClientset,
+	wf *factory.WatchFactory, recorder record.EventRecorder) *secondaryNetworkClusterManager {
+	klog.Infof("Creating new multi network cluster manager")
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	sncm := &secondaryNetworkClusterManager{
+		ovnClient:    ovnClient,
+		client:       ovnClient.KubeClient,
+		kube:         kube,
+		watchFactory: wf,
+		stopChan:     make(chan struct{}),
+	}
+	sncm.nadController = nad.NewNetAttachDefinitionController(
+		sncm, ovnClient.NetworkAttchDefClient, recorder)
+	return sncm
+}
+
+// Start starts the secondary layer3 controller, handles all events and creates all needed logical entities
+func (sncm *secondaryNetworkClusterManager) Start() error {
+	klog.Infof("Starting net-attach-def controller")
+	return sncm.nadController.Run(sncm.stopChan)
+}
+
+func (sncm *secondaryNetworkClusterManager) Stop() {
+	close(sncm.stopChan)
+
+	// stops each network controller associated with net-attach-def; it is ok
+	// to call GetAllControllers here as net-attach-def controller has been stopped,
+	// and no more change of network controllers
+	klog.Infof("Stops net-attach-def controller")
+	for _, oc := range sncm.nadController.GetAllNetworkControllers() {
+		oc.Stop()
+	}
+}
+
+// NewNetworkController implements the networkAttachDefController.NetworkControllerManager
+// interface function.  This function is called by the net-attach-def controller when
+// a layer2 or layer3 secondary network is created.  Layer2 type is not handled here.
+func (sncm *secondaryNetworkClusterManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (nad.NetworkController, error) {
+	klog.Infof("New net-attach-def controller for network %s called", nInfo.GetNetworkName())
+	topoType := netConfInfo.TopologyType()
+	if topoType == ovntypes.Layer3Topology {
+		sncc := newSecondaryNetworkClusterController(sncm.ovnClient, sncm.watchFactory,
+			nInfo, netConfInfo, nInfo.GetNetworkName())
+		return sncc, nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
```

Comment:
Maybe we should handle the other topologies that don't need subnet allocation with something different than a generic error? Otherwise the nad controller would just be retrying?

Reply:
If we don't return error, then net-attach-def-controller will store the returned oc  here - https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/network-controller-manager/network_attach_def_controller.go#L462

So how do we handle it ?  Return a dummy controller ?   Or perhaps netAttachDefinitionController should not even create a controller if the secondary network controller type is not layer 3 ?

Reply:
I guess we are introducing a semantic that was not there before to the nad controller that not all networks need a network manager. 

Options:
* Return a specific error which the nad controller could manage specifically
* Pass a predicate to the nad controller that would evaluate which networks need a manager
* Return a noop manager

Reply:
Actually, we would need to handle and update flow where a network which had a network manager no longer needed it :thinking:

Reply:
Ah forget about the update flow concerns. The nad controller deletes the old manager and creates a new one on update.